### PR TITLE
[Feature] Initial concept for adding percentage based goals

### PR DESCRIPTION
**Pull request by shall0pass**
This is an initial concept for adding percent based goal targets.

This version works by using a string in the form of: 
#template 10% of Income   <- Income is an income category name.  Only income category names will work here currently.
or
#template 10% of all income<- 'all income' is a keyword in this context and will base the calculation on the total income for the month.

Some of the nicer touches like Jed's polite notification that the syntax isn't correct is not implemented here yet.

### CODE REVIEW

**Comment by youngcw** [comment]
Would a check for if the income category has a negative amount be worth while? or is that up to the user to not mess up... like I did just now?

A simple test of this is looking great!


Are there plans to implement features similar to the existing "$X every Y weeks" options of the other templates?  That could probably cover the need to overwrite the value if your get paid weekly, bi-weekly, or inconsistently.

File: packages/loot-core/src/server/budget/goaltemplates.js [hunk 1]
```diff
@@ -329,6 +329,24 @@ async function applyCategoryTemplate(category, template_lines, month, force) {
         break;
       }
       case 'percentage': {
+        /*expected behavior:  
+          By using the string '#template 10% of Income' , the following code should calculate 10% of the 'Income' category and budget the appropriate amount.
+          By using the string '#template 10% of IncomeGroup', the keyword 'IncomeGroup' should sum all available income within the 'Income' category group, 
+        */
+        let percent = template.percent;
+        let monthlyIncome = 0;
+        if (template.category == 'IncomeGroup') {  //adds a keyword 'IncomeGroup' to use all income for the month for the calculation
```
[the remark this answers is not in the document]
Comment (shall0pass) [reply packages/loot-core/src/server/budget/goaltemplates.js:338]: I like that a lot better.  Done.

File: packages/loot-core/src/server/budget/goaltemplates.js [hunk 2]
```diff
@@ -329,6 +329,24 @@ async function applyCategoryTemplate(category, template_lines, month, force) {
...
+          monthlyIncome = await getSheetValue(sheetName, `total-income`);
+        }
+        else { 
+          let income_category = (await db.getCategories()).filter(c => c.is_income == true && c.name == template.category);
```
[the remark this answers is not in the document]
Comment (shall0pass) [reply packages/loot-core/src/server/budget/goaltemplates.js:342]: I have an error occuring when using the find() function.

[Exception] TypeError: Cannot read properties of undefined (reading 'id')
    at applyCategoryTemplate (goaltemplates.js:345:46)
    at async processTemplate (goaltemplates.js:57:25)

**Comment by shall0pass** [comment]
> Would a check for if the income category has a negative amount be worth while? or is that up to the user to not mess up... like I did just now?

That's a good idea.  I'll add a basic check so the budgeted amount can't be negative.  If a negative amount is entered into the income section, the all income calculation could be affected.  I'm not sure how to check for that.
> 
> A simple test of this is looking great!
> 
Thanks.

> Are there plans to implement features similar to the existing "$X every Y weeks" options of the other templates? That could probably cover the need to overwrite the value if your get paid weekly, bi-weekly, or inconsistently.

I don't know that any specific plans are in place.  If you could describe a specific template string, I would strongly reccomend making an issue so we can track it and someone could try to pick it up.

File: packages/loot-core/src/server/budget/goaltemplates.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (j-f1) [reply packages/loot-core/src/server/budget/goaltemplates.js]: `find` will find exactly one object and return it while `filter` returns an array of matching objects so you’ll need to remove the `[0]` below.

File: packages/loot-core/src/server/budget/goaltemplates.js [hunk 3]
```diff
@@ -329,19 +329,25 @@ async function applyCategoryTemplate(category, template_lines, month, force) {
...
           }
-          */
+          monthlyIncome = await getSheetValue(
+            sheetName,
+            `sum-amount-${income_category[0].id}`,
+          );
+        }
+        if (monthlyIncome > 0) {
+          to_budget = Math.round(monthlyIncome * (percent / 100));
+        } else to_budget = 0;
```
Comment (j-f1) [inline packages/loot-core/src/server/budget/goaltemplates.js:350]: Suggestion: either replace this with a call to `Math.max`
```suggestion
        to_budget = Math.max(0, Math.round(monthlyIncome * (percent / 100)));
```

…or add braces to the `else`
```suggestion
        if (monthlyIncome > 0) {
          to_budget = Math.round(monthlyIncome * (percent / 100));
        } else {
          to_budget = 0;
        }
```

File: packages/loot-core/src/server/budget/goaltemplates.js [hunk 4]
```diff
@@ -329,19 +329,25 @@ async function applyCategoryTemplate(category, template_lines, month, force) {
...
-          for (var g = 0; g < budget.categoryGroups.length; g++) {
-            if (income_category.group_id == budget.categoryGroups[g].id) {
-              for (var c = 0; c < budget.categoryGroups[g].categories.length; c++)
-                if (income_category.id == budget.categoryGroups[g].categories[c].id) {
-                  let month_category = budget.categoryGroups[g].categories[c];
-                }
-            }
+        let percent = template.percent;
+        let monthlyIncome = 0;
+        if (template.category === 'all income') {
```
Comment (j-f1) [inline packages/loot-core/src/server/budget/goaltemplates.js:334]: the language as a whole is case insensitive, no reason to be different here (right?)

```suggestion
        if (template.category.toLowerCase() === 'all income') {
```

File: packages/loot-core/src/server/budget/goaltemplates.js [hunk 5]
```diff
@@ -329,19 +329,25 @@ async function applyCategoryTemplate(category, template_lines, month, force) {
...
+          monthlyIncome = await getSheetValue(sheetName, `total-income`);
+        } else {
+          let income_category = (await db.getCategories()).filter(
+            c => c.is_income == true && c.name == template.category,
```
Comment (j-f1) [inline packages/loot-core/src/server/budget/goaltemplates.js:338]:
```suggestion
            c => c.is_income == true && c.name.toLowerCase() == template.category.toLowerCase(),
```

File: packages/loot-core/src/server/budget/goaltemplates.js [hunk 6]
```diff
@@ -329,19 +329,25 @@ async function applyCategoryTemplate(category, template_lines, month, force) {
...
+          );
+          if (!income_category) {
+            throw new Error(`Could not find category “${template.category}”`);
```
Comment (j-f1) [inline packages/loot-core/src/server/budget/goaltemplates.js:341]: Also I know I suggested adding this but we currently don’t handle errors well here. We should update the `applyCategoryTemplate` API to return a value or an error, and then update all the places where `console.log` is currently called to return an error too. This should probably be done in a different PR that will merge before this one, though.

Comment (j-f1) [reply packages/loot-core/src/server/budget/goaltemplates.js:341]: ⏩ 859

File: packages/loot-core/src/server/budget/goaltemplates.js (hunk 3 above)
Comment (shall0pass) [reply packages/loot-core/src/server/budget/goaltemplates.js:350]: done.

File: packages/loot-core/src/server/budget/goaltemplates.js (hunk 6 above)
Comment (j-f1) [reply packages/loot-core/src/server/budget/goaltemplates.js:319]: Can you update this to work with the new error logic?

File: packages/loot-core/src/server/budget/goaltemplates.js [hunk 7]
```diff
@@ -305,19 +305,25 @@ async function applyCategoryTemplate(category, template_lines, month, force) {
...
-                }
-            }
+        let percent = template.percent;
+        let monthlyIncome = 0;
+        if (template.category.toLowerCase() === 'all income') {
+          monthlyIncome = await getSheetValue(sheetName, `total-income`);
+        } else {
+          let income_category = (await db.getCategories()).find(
+            c =>
+              c.is_income === 1 &&
```
Comment (j-f1) [inline packages/loot-core/src/server/budget/goaltemplates.js:315]: also small nit: this seems to be the convention used by other code that checks this property:
```suggestion
              c.is_income &&
```

File: packages/loot-core/src/server/budget/goaltemplates.js (hunk 6 above)
Comment (shall0pass) [reply packages/loot-core/src/server/budget/goaltemplates.js:319]: I will.  I merged the master in already, so I'm just waiting to get back to a PC.

**Review by j-f1 — APPROVED** [review]
